### PR TITLE
[DROOLS-3742] Implement Test scenario coverage report for DMN

### PR DESCRIPTION
**Pull request by danielezonca**
This PR implement both backend and frontend of test coverage report for DMN.

https://issues.jboss.org/browse/DROOLS-3742

NOTE: some of the features described in the mockups are related to other tickets not implemented yet  so the UI is a bit different ( http://bpam-frontend.surge.sh/drools-test-editor-coverage.html )
- no `View diagram` https://issues.jboss.org/browse/DROOLS-3390
- no `Download report` https://issues.jboss.org/browse/DROOLS-3486
- no `View audit log` https://issues.jboss.org/browse/DROOLS-3486
- no `Decision table button` https://issues.jboss.org/browse/DROOLS-3390
- no Rule support https://issues.jboss.org/browse/DROOLS-3743

@kkufova @gitgabrio @Rikkola

### CODE REVIEW

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/CoverageReportDonutPresenter.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+import org.dashbuilder.dataset.DataSetFactory;
+import org.dashbuilder.displayer.DisplayerSettingsFactory;
+import org.dashbuilder.displayer.client.Displayer;
+import org.dashbuilder.displayer.client.DisplayerCoordinator;
+import org.dashbuilder.displayer.client.DisplayerLocator;
+import org.drools.workbench.screens.scenariosimulation.client.resources.i18n.ScenarioSimulationEditorConstants;
+import org.jboss.errai.common.client.dom.elemental2.Elemental2DomUtil;
+
+@Dependent
+public class CoverageReportDonutPresenter {
```
Comment (danielezonca) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/CoverageReportDonutPresenter.java]: Thanks @Rikkola for `TestResultDonutPresenter`. This class (and the test one) is a sort of clone :P

**Comment by kkufova** [comment]
@danielezonca, could you please rebase this on master? Recently, a number of changes have been merged (https://github.com/kiegroup/drools-wb/pull/1095) that are causing the local build to fail.

**Comment by danielezonca** [comment]
@kkufova Done

**Comment by danielezonca** [comment]
jenkins retest this please

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/CoverageReportViewImpl.html [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+                <div id="kieCoverageChart1" data-field="donutChart" class="example-donut-chart-utilization"></div>
+                <!------- END OF DONUT CHART -->
+
+            </div>
+        </div>
+
+        <div class="row" data-field="decisionListSection">
+            <div class="col-xs-12 col-md-12">
+                <hr class="kie-dock__divider"/>
+                Number of times a Decision has been evaluated
```
Comment (kkufova) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/CoverageReportViewImpl.html]: Please change to:

> A number of times each decision has been evaluated:

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties [hunk 3]
```diff
@@ -127,9 +127,15 @@ removeCollectionQuestion=Do you want to continue and remove the collection?
 removeCollectionWarningText=This action cannot be undone.
 #
 settings=Settings
+coverageReport=Coverage Report
 uploadWarning=Make sure the headers in the selected CSV file have not been modified. Otherwise, the spreadsheet may not be successfully imported.
 importFailedMessage=A CSV file with an incorrect header layout has been imported. Make sure the imported CSV file has unmodified headers.
 importErrorTitle=Error during import
+executed=Executed
+notCovered=notCovered
+decisionsEvaluated=Decision evaluated for Scenario
```
Comment (kkufova) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties]: Please change to:

> Decisions evaluated for test scenario

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties [hunk 4]
```diff
@@ -127,9 +127,15 @@ removeCollectionQuestion=Do you want to continue and remove the collection?
 removeCollectionWarningText=This action cannot be undone.
 #
 settings=Settings
+coverageReport=Coverage Report
 uploadWarning=Make sure the headers in the selected CSV file have not been modified. Otherwise, the spreadsheet may not be successfully imported.
 importFailedMessage=A CSV file with an incorrect header layout has been imported. Make sure the imported CSV file has unmodified headers.
 importErrorTitle=Error during import
+executed=Executed
+notCovered=notCovered
```
Comment (kkufova) [inline drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties]: I would prefer to use a single word there. Could we for example use _Missed_, _Ignored_, or _Omitted_ (this one is my favorite)?

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties [hunk 5]
```diff
@@ -127,9 +127,15 @@ removeCollectionQuestion=Do you want to continue and remove the collection?
...
+decisionsEvaluated=Decision evaluated for Scenario
+coverageNotSupportedForRule=Coverage report is not supported for Rule based scenarios
```
[the remark this answers is not in the document]
Comment (danielezonca) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties]: Done

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties [hunk 6]
```diff
@@ -127,9 +127,15 @@ removeCollectionQuestion=Do you want to continue and remove the collection?
...
+runATestToSeeCoverageReport=Run a scenario to see coverage report
```
[the remark this answers is not in the document]
Comment (danielezonca) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties]: Done

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties (hunk 3 above)
Comment (danielezonca) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties]: Done

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties (hunk 4 above)
Comment (danielezonca) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/resources/org/drools/workbench/screens/scenariosimulation/client/resources/i18n/ScenarioSimulationEditorConstants.properties]: Done (used Omitted)

File: drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/CoverageReportViewImpl.html (hunk 2 above)
Comment (danielezonca) [reply drools-wb-screens/drools-wb-scenario-simulation-editor/drools-wb-scenario-simulation-editor-client/src/main/java/org/drools/workbench/screens/scenariosimulation/client/rightpanel/CoverageReportViewImpl.html]: Done

**Comment by danielezonca** [comment]
> @danielezonca, I have a few of notes:
> 
> 1. First of all, I would like to clarify how exactly the coverage report should work. For example, [this](https://issues.jboss.org/secure/attachment/12450212/Traffic%20Violation.dmn) DMN file contains two decisions; one of them is a decision table (with 4 rows) and the other one is a context. I assumed that the coverage report will take into account all the information available **in** the decisions; that is, all the rows available in the decision table (in case of the first decision). For example, see the following screenshot:
> 
> bug1
> 
> As you can see, only one row of the decision table is covered, however, the coverage is 100%. Yes, we covered both decisions, but the actual test coverage is questionable to say the least. I wonder whether we could somehow take into account the information inside decisions as well. However, if this is how it should work, let me know.

Coverage is at decision level, not at single row of a decision table so it is enough that a certain decision has been executed to be considered covered (see comment https://issues.jboss.org/browse/DROOLS-3742?focusedCommentId=13722541&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-13722541 and reply).
This means for instance that if you just specify a correct input for each input node, the coverage of the decision will be 100% even if none of the decisions has been verified. The same concept will be implemented for rules: if a rule has been executed it means that all the preconditions are satisfied and this is enough.
The general idea is not to have a code coverage (= cover all the branches) but execute all the decisions successfully given the inputs.

> 
> 2. Secondly, I have a couple of UX notes:
> 
> * As can be seen in the screenshot above, two captions are not displayed completely (**Percentage of decisions t...** and **Should the driver be susp...**). They are not uncollapsed even after resizing the panel. I think these should always be displayed as a whole, even if it would mean that they will take two lines instead of one. Can this be fixed?

@srambach can you please update your HTML ( http://bpam-frontend.surge.sh/drools-test-editor-coverage.html#0 ) as described?

> * I think the pie chart should be bigger. It is almost impossible to read the description after a mouseover.

@Rikkola I'm using the same donut of test result, do you have any idea about how to increase it?

> * There should not be a space between a number and a percentage sign in English language.

Done

> * I added a few of comments to change some captions.

Done

**Comment by Rikkola** [comment]
@danielezonca @kkufova I think the chart just fills the remaining place. Based on my research the CSS settings for the donut are hidden behind the API, but could be there is a way. Might be this needs to be checked from the Appformer team.

**Comment by danielezonca** [comment]
@kkufova 
CSS fixes done
Created https://issues.jboss.org/browse/DROOLS-3975 for donut

**Comment by danielezonca** [comment]
@manstis 
Can you please squash and merge this?

**Pull request merged**